### DISCUSSION TITLE
How to prevent race conditions for following code?

### DISCUSSION THREAD

**Discussion by wuarmin** [jeremyevans/sequel#2024] [General]
Hey,

I experience race conditions (threads, processes) with following code. Hint: The `insert` and the `find` have to be separate calls. I cannot use `returning`.

```ruby
def upsert(data)
  id = DB[:users].insert_conflict(target: :id, update: {
    name: Sequel[:excluded][:name],
    email: Sequel[:excluded][:email],
    updated_at: Sequel[:excluded][:updated_at],
   }).insert(data)
   
 DB[:users][id: id]
end
```

If multiple threads/processes call this code, it sometimes fails with following invalid sql-statement:

```sql
INSERT INTO "users" ("id", "name", "email", "created_at", "updated_at")
VALUES (2, 'Test', 'test@test.at', '2023-04-17 16:38:11.512975+0200', '2023-04-17 16:38:11.512975+0200')
ON CONFLICT ("id")
DO UPDATE SET "name" = "excluded"."name", "email" = "excluded"."email", "updated_at" = "excluded"."updated_at"
RETURNING "2" -- here's the failure (it should be RETURNING "id", column "2" does not exist)
```

A db-transaction obviously doesn't help. Do I need to sync the overlying code (i.e. Redlock (redis process-lock))?

Thanks! 
best regards

**Comment by jeremyevans** [comment]
What database are you using?  Does  switching `DB[:users].returning(:id).insert_conflict` fix the issue?  Is it possible for you to put together a self contained example showing the issue?

**Reply by wuarmin** [reply]
I'm using postgres:15.0. 
I'll try it. Do you have a guess?

**Reply by jeremyevans** [reply]
Only guess would be you have simultaneous access to the same connection object, such as non disconnecting before fork when using multiprocess code, or using the single threaded connection pool in a multithreaded application.  Both of those would be a bug in the application and not in Sequel.  I couldn't rule out a bug in Sequel at this point, but nothing comes to mind as a possible cause.

**Reply by wuarmin** [reply]
Thanks for your input. If I change the code to
```ruby
def upsert(data)
  User.new(DB[:users].returning(:id, :name, :email, :updated_at, :created_at).insert_conflict(target: :id, update: {
    name: Sequel[:excluded][:name],
    email: Sequel[:excluded][:email],
    updated_at: Sequel[:excluded][:updated_at],
   }).insert(data).first)
end
```
the issue seems to be fixed.

Following code-change does not fix the issue:

```ruby
def upsert(data)
  id = DB[:users].returning(:id).insert_conflict(target: :id, update: {
    name: Sequel[:excluded][:name],
    email: Sequel[:excluded][:email],
    updated_at: Sequel[:excluded][:updated_at],
   }).insert(data).first[:id])

  DB[:users][id: id]
end
```

**Reply by jeremyevans** [reply]
You probably want `User.call` or `User.load` instead of `User.new`, since these are for users that already exist in the database.

**Reply by wuarmin** [reply]
Thank you @jeremyevans. 
With...

> such as non disconnecting before fork when using multiprocess code

...you put me on the right track. 
I think I found the issue(https://github.com/decafsucks/decafsucks/pull/19). What do you think?

**Reply by jeremyevans** [reply]
That could certainly be the cause of the issue.  When using `fork` for multiprocess code with Sequel, you should always make sure database connections are disconnected before fork. Sequel will reestablish connections as needed in the child processes.
